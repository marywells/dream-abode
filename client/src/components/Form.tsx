import { useState } from 'react';
import { IProperty } from '../types';

interface IProps {
  addNewProperty: (property: IProperty) => void;
}
export function Form({ addNewProperty }: IProps) {
  const [name, setName] = useState('');
  const [imageURL, setImageURL] = useState('');
  const [description, setDescription] = useState('');
  const [price, setPrice] = useState(0);
  const [beds, setBeds] = useState(0);
  const [baths, setBaths] = useState(0);
  const [type, setType] = useState('');

  function createProperty(newProperty: IProperty) {
    addNewProperty(newProperty);
  }

  function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    const property = {
      name,
      imageURL,
      description,
      price,
      beds,
      baths,
      type,
    };

    createProperty(property);

    setName('');
    setImageURL('');
    setDescription('');
    setPrice(0);
    setBeds(0);
    setBaths(0);
    setType('');
  }

  return (
    <div className={styles.form}>
      <h1 className={styles.formHeader}>Add a Property</h1>
      <form onSubmit={handleSubmit}>
        <div className={styles.inputLabel}>Name</div>
        <input
          className={styles.input}
          type='text'
          placeholder='First line of address'
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
        <div className={styles.inputLabel}>Image</div>
        <input
          className={styles.input}
          type='text'
          placeholder='Image URL'
          value={imageURL}
          onChange={(e) => setImageURL(e.target.value)}
        />
        <div className={styles.inputLabel}>Description</div>
        <input
          className={styles.input}
          type='text'
          placeholder='Short description'
          value={description}
          onChange={(e) => setDescription(e.target.value)}
        />
        <div className={styles.inputLabel}>Price</div>
        <input
          className={styles.input}
          type='number'
          placeholder='Set price'
          value={price}
          onChange={(e) => setPrice(e.target.valueAsNumber)}
        />
        <div className={styles.inputLabel}>Bedrooms</div>
        <input
          className={styles.input}
          type='number'
          placeholder='Set bedrooms'
          value={beds}
          onChange={(e) => setBeds(e.target.valueAsNumber)}
        />
        <div className={styles.inputLabel}>Bathrooms</div>
        <input
          className={styles.input}
          type='number'
          placeholder='Set bathrooms'
          value={baths}
          onChange={(e) => setBaths(e.target.valueAsNumber)}
        />
        <div className={styles.inputLabel}>Property Type</div>
        <select
          className={styles.input}
          value={type}
          onChange={(e) => setType(e.target.value)}
        >
          <option value='Terraced'>Terraced</option>
          <option value='Flat'>Flat</option>
          <option value='Detached'>Detached</option>
          <option value='Detached'>Semi-Detached</option>
          <option value='Castle'>Castle</option>
        </select>
        <br />
        <input className={styles.button} type='submit' value='Create' />
      </form>
    </div>
  );
}

const styles = {
  form: 'w-72 md:fixed md:top-20 md:right-2 bg-gray-800 border p-4 m-2 rounded-lg',
  formHeader: 'text-white text-xl font-bold ',
  inputLabel: 'text-white text-xs pt-2',
  input: 'rounded text-sm w-full',
  button:
    'bg-transparent text-white font-semibold hover:text-white border mt-2 py-1 px-3 rounded-full',
};
