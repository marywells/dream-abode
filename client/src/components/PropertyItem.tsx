import { IProperty } from '../types';

export function PropertyItem(props: IProperty) {
  const { name, type, description, imageURL, price, beds, baths } = props;
  return (
    <div className={styles.propertyContainer}>
      <div data-testid='propertyItemName' className={styles.nameTile}>
        <p className={styles.name}>{name}</p>
      </div>

      <div className={styles.propertyDetails}>
        <div data-testid='propertyItemPrice' className={styles.price}>
          £{price?.toLocaleString('en-US')}
        </div>
        <div className={styles.type}>{type}</div>
        <div data-testid='propertyItemBeds'>{beds} bedrooms</div>
        <div>{baths} bathrooms</div>
        <br />
      </div>
      <div className={styles.imageContainer}>
        <img className={styles.image} src={imageURL} alt={name} />
      </div>
      <div className={styles.description}>{description}</div>
      <div className={styles.editDelete}>
        <button>✏️</button>
        <button>❌</button>
      </div>
    </div>
  );
}

const styles = {
  propertyContainer:
    'flex-wrap relative w-full border-2 rounded-lg shadow m-6 ml-2 text-xs flex',
  nameTile: 'flex bg-gray-900 rounded-lg items-center text-center',
  name: 'w-40 text-lg p-2 text-white ',
  imageContainer: 'w-40',
  image: 'h-28 rounded-lg',
  propertyDetails: 'w-24 pl-4 pt-2 mr-2',
  price: 'text-lg font-bold',
  type: 'font-semibold',
  description: 'hidden md:block w-72 p-2',
  editDelete: 'absolute bottom-2 right-2',
};
