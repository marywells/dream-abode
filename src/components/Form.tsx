export function Form() {
  return <div>Add a Property Form</div>;
}

const styles = {};
