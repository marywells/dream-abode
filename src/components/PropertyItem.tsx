import { IProperty } from '../types';

export function PropertyItem(props: IProperty) {
  return (
    <div className={styles.propertyContainer}>
      <p>{props.name}</p>
      <p>{props.type}</p>
      <p>{props.description}</p>
      <p>Â£{props.price}</p>
      <p>{props.beds} bedrooms</p>
      <p>{props.baths} bathrooms</p>
    </div>
  );
}

const styles = {
  propertyContainer: 'border-2 rounded-lg text-xs shadow m-2 p-1',
};
