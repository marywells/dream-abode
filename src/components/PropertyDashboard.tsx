import { useState, useEffect } from 'react';
import { PropertyItem } from './PropertyItem';
import { Form } from './Form';
import * as ApiService from '../api-service';
import { IProperty } from '../types';

export function PropertyDashboard() {
  const [properties, setProperties] = useState<IProperty[]>([]);

  useEffect(() => {
    loadProperties();
  }, []);

  async function loadProperties() {
    await ApiService.getAll().then((data) => {
      setProperties(data);
    });
  }

  function addNewProperty(property: IProperty) {
    ApiService.postOne(property).then(() => loadProperties());
  }

  return (
    <div className={styles.dashboardContainer}>
      <div className={styles.propertyList}>
        <h1>Property Dashboard</h1>
        {properties.map((property) => {
          return (
            <div key={property._id}>
              {
                <PropertyItem
                  _id={property._id}
                  name={property.name}
                  imageURL={property.imageURL}
                  description={property.description}
                  price={property.price}
                  beds={property.beds}
                  baths={property.baths}
                  type={property.type}
                />
              }
            </div>
          );
        })}
      </div>

      <Form addNewProperty={addNewProperty} />
    </div>
  );
}

const styles = {
  dashboardContainer: 'relative pt-16',
  propertyList: 'absolute flex-col',
};
