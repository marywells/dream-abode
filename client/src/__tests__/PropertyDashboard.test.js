import ReactDOM from 'react-dom';
import { PropertyDashboard } from '../components/PropertyDashboard';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';

it('renders property dashboard on the page', () => {
  const div = document.createElement('div');
  ReactDOM.render(<PropertyDashboard />, div);
});

it('matches property dashboard snapshot', () => {
  const tree = renderer.create(<PropertyDashboard />).toJSON();
  expect(tree).toMatchSnapshot();
});
