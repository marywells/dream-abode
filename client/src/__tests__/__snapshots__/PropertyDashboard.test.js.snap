// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches property dashboard snapshot 1`] = `
<div
  className="flex-wrap md:relative p-6 pt-20"
>
  <div
    className="md:w-3/5 flex-col"
  >
    <h1
      className="font-bold text-gray-900 pt-2"
    >
      “Perhaps home is not a place but simply an irrevocable condition.” James Baldwin
    </h1>
  </div>
  <div
    className="w-72 md:fixed md:top-20 md:right-2 bg-gray-800 border p-4 m-2 rounded-lg"
  >
    <h1
      className="text-white text-xl font-bold "
    >
      Add a Property
    </h1>
    <form
      onSubmit={[Function]}
    >
      <div
        className="text-white text-xs pt-2"
      >
        Name
      </div>
      <input
        className="rounded text-sm w-full"
        onChange={[Function]}
        placeholder="First line of address"
        type="text"
        value=""
      />
      <div
        className="text-white text-xs pt-2"
      >
        Image
      </div>
      <input
        className="rounded text-sm w-full"
        onChange={[Function]}
        placeholder="Image URL"
        type="text"
        value=""
      />
      <div
        className="text-white text-xs pt-2"
      >
        Description
      </div>
      <input
        className="rounded text-sm w-full"
        onChange={[Function]}
        placeholder="Short description"
        type="text"
        value=""
      />
      <div
        className="text-white text-xs pt-2"
      >
        Price
      </div>
      <input
        className="rounded text-sm w-full"
        onChange={[Function]}
        placeholder="Set price"
        type="number"
        value={0}
      />
      <div
        className="text-white text-xs pt-2"
      >
        Bedrooms
      </div>
      <input
        className="rounded text-sm w-full"
        onChange={[Function]}
        placeholder="Set bedrooms"
        type="number"
        value={0}
      />
      <div
        className="text-white text-xs pt-2"
      >
        Bathrooms
      </div>
      <input
        className="rounded text-sm w-full"
        onChange={[Function]}
        placeholder="Set bathrooms"
        type="number"
        value={0}
      />
      <div
        className="text-white text-xs pt-2"
      >
        Property Type
      </div>
      <select
        className="rounded text-sm w-full"
        onChange={[Function]}
        value=""
      >
        <option
          value="Terraced"
        >
          Terraced
        </option>
        <option
          value="Flat"
        >
          Flat
        </option>
        <option
          value="Detached"
        >
          Detached
        </option>
        <option
          value="Detached"
        >
          Semi-Detached
        </option>
        <option
          value="Castle"
        >
          Castle
        </option>
      </select>
      <br />
      <input
        className="bg-transparent text-white font-semibold hover:text-white border mt-2 py-1 px-3 rounded-full"
        type="submit"
        value="Create"
      />
    </form>
  </div>
</div>
`;
